File: cprdspy/CPR_js2/Spirals/spiral.js
/**
 * CPR 图形库 - JavaScript 版本 (Plotly.js)
 * Spirals 模块 - 螺旋线相关函数
 */

/**
 * 绘制对数螺线
 */
function logSpiral({
    n = 4,
    a = 1,
    b = 1,
    cyc = 0.5,
    color = "b",
    theta = 0,
    rotation = 0,
    direction = "both",
    alpha = 1,
    lineWidth = 1,
} = {}) {
    if (!["both", "in", "out"].includes(direction)) {
        throw new Error("direction must be 'both', 'in' or 'out'");
    }

    let tStart, tEnd;
    if (direction === "both") {
        tStart = -cyc * 2 * Math.PI;
        tEnd = cyc * 2 * Math.PI;
    } else if (direction === "out") {
        tStart = 0;
        tEnd = cyc * 2 * Math.PI;
    } else {
        tStart = 0;
        tEnd = -cyc * 2 * Math.PI;
    }

    const points = 1000;
    const t = Array.from({ length: points + 1 }, (_, i) => tStart + ((tEnd - tStart) * i) / points);
    const r = t.map(t => Math.pow(Math.cos(Math.PI / n), (-n * t) / Math.PI));
    const ang = t.map(t => t + theta + rotation);
    const x = r.map((r, i) => a * r * Math.cos(ang[i]));
    const y = r.map((r, i) => b * r * Math.sin(ang[i]));

    return {
        type: "scattergl",
        mode: "lines",
        x: x,
        y: y,
        line: { color: color, width: lineWidth },
        opacity: alpha,
        hoverinfo: "none",
    };
}

/**
 * 绘制 n 瓣螺旋
 */
function nSpiral({
    n = 4,
    N = 4,
    cyc = 0.5,
    color = "b",
    theta = 0,
    rotation = 0,
    lineWidth = 1,
} = {}) {
    const traces = [];

    for (let i = 0; i < N; i++) {
        const t1 = logSpiral({
            n, a: 1, b: 1, cyc, color,
            theta: theta + (i * 2 * Math.PI) / N,
            rotation, direction: "both", lineWidth,
        });

        const t2 = logSpiral({
            n, a: -1, b: 1, cyc, color,
            theta: -theta + (i * 2 * Math.PI) / N,
            rotation: -rotation, direction: "both", lineWidth,
        });

        traces.push(t1, t2);
    }

    return traces;
}

/**
 * 双向对数螺线
 */
function logSpiralInOut({ n, a, b, cyc, color = "b", theta = 0 } = {}) {
    const points = 1000;
    const tStart = -cyc * 2 * Math.PI;
    const tEnd = cyc * 2 * Math.PI;
    const t = Array.from({ length: points + 1 }, (_, i) => tStart + ((tEnd - tStart) * i) / points);
    const r = t.map(t => Math.pow(Math.cos(Math.PI / n), (-n * t) / Math.PI));
    const x = r.map((r, i) => a * r * Math.cos(t[i] + theta));
    const y = r.map((r, i) => b * r * Math.sin(t[i] + theta));

    return {
        type: "scattergl",
        mode: "lines",
        x: x,
        y: y,
        line: { color: color },
        hoverinfo: "none",
    };
}

/**
 * 外向对数螺线
 */
function logSpiralOut({ n, a, b, cyc, color = "b", theta = 0 } = {}) {
    const points = 100;
    const t = Array.from({ length: points + 1 }, (_, i) => (cyc * 2 * Math.PI * i) / points);
    const r = t.map(t => Math.pow(Math.cos(Math.PI / n), (-n * t) / Math.PI));
    const x = r.map((r, i) => a * r * Math.cos(t[i] + theta));
    const y = r.map((r, i) => b * r * Math.sin(t[i] + theta));

    return {
        type: "scattergl",
        mode: "lines",
        x: x,
        y: y,
        line: { color: color },
        hoverinfo: "none",
    };
}

/**
 * 内向对数螺线
 */
function logSpiralIn({ n, a, b, cyc, color = "b", theta = 0 } = {}) {
    const points = 100;
    const t = Array.from({ length: points + 1 }, (_, i) => (-cyc * 2 * Math.PI * i) / points);
    const r = t.map(t => Math.pow(Math.cos(Math.PI / n), (-n * t) / Math.PI));
    const x = r.map((r, i) => a * r * Math.cos(t[i] + theta));
    const y = r.map((r, i) => b * r * Math.sin(t[i] + theta));

    return {
        type: "scattergl",
        mode: "lines",
        x: x,
        y: y,
        line: { color: color },
        hoverinfo: "none",
    };
}

/**
 * 绘制 n 个螺旋
 */
function nSpiralRotate({ n, cyc, color, alpha = 0, theta = 0 } = {}) {
    const traces = [];

    for (let i = 0; i < n; i++) {
        const t1 = logSpiral({
            n, a: 1, b: 1, cyc, color,
            theta: alpha + theta + (i * 2 * Math.PI) / n,
            lineWidth: 1,
        });

        const t2 = logSpiral({
            n, a: -1, b: 1, cyc, color,
            theta: alpha - theta + (i * 2 * Math.PI) / n,
            lineWidth: 1,
        });

        traces.push(t1, t2);
    }

    return traces;
}

/**
 * 旋转外向螺旋
 */
function nSpiralRotateOut({ n, cyc, color, theta = 0 } = {}) {
    const traces = [];

    for (let i = 0; i < n; i++) {
        const t1 = logSpiralOut({ n, a: 1, b: 1, cyc, color, theta: theta + (i * 2 * Math.PI) / n });
        const t2 = logSpiralOut({ n, a: -1, b: 1, cyc, color, theta: -theta + (i * 2 * Math.PI) / n });
        traces.push(t1, t2);
    }

    return traces;
}

/**
 * 旋转内向螺旋
 */
function nSpiralRotateIn({ n, cyc, color, theta = 0 } = {}) {
    const traces = [];

    for (let i = 0; i < n; i++) {
        const t1 = logSpiralIn({ n, a: 1, b: 1, cyc, color, theta: theta + (i * 2 * Math.PI) / n });
        const t2 = logSpiralIn({ n, a: -1, b: 1, cyc, color, theta: -theta + (i * 2 * Math.PI) / n });
        traces.push(t1, t2);
    }

    return traces;
}

/**
 * 马蹄莲花瓣
 */
function callaPetal({ n, cyc, theta, color } = {}) {
    const t1 = logSpiral({ n, a: 1, b: 1, cyc: cyc * 1.25, color, theta, lineWidth: 1 });
    const t2 = logSpiral({ n, a: -1, b: 1, cyc: cyc * 1.25, color, theta: -theta, lineWidth: 1 });
    return [t1, t2];
}

/**
 * 多个马蹄莲花瓣
 */
function callaByPetal({ n, cyc, N, theta, colors } = {}) {
    const traces = [];

    for (let i = 0; i < N; i++) {
        const petalColor = colors[i % colors.length];
        const petals = callaPetal({
            n, cyc,
            theta: theta + (i * 2 * Math.PI) / N,
            color: petalColor,
        });
        traces.push(...petals);
    }

    return traces;
}

export {
    logSpiral,
    logSpiralOut,
    logSpiralIn,
    logSpiralInOut,
    nSpiral,
    nSpiralRotate,
    nSpiralRotateOut,
    nSpiralRotateIn,
    callaPetal,
    callaByPetal,
};

File: cprdspy/CPR_plotly1/Spirals/spiral.py
import plotly.graph_objects as go
import numpy as np


def logSpiral(
    n=4,
    a=1,
    b=1,
    cyc=0.5,
    color="b",
    theta=0,
    rotation=0,
    direction="both",
    alpha=1,
    line_width=1.0,
    plot=True,
):
    """绘制广义对数螺线。

    参数说明：
    - n: 花瓣/对称数（影响基底 cos(pi/n)）
    - a, b: x/y 方向缩放（可做椭圆化）
    - cyc: 周期数（每个周期为 2π）
    - color: 线颜色
    - theta: 全局相位偏移
    - rotation: 额外角度偏移
    - direction: 'both'|'in'|'out' 控制绘制方向
    - alpha: 透明度
    - line_width: 线宽
    - plot: 是否返回 go.Scatter 轨迹
    """
    if direction not in ("both", "in", "out"):
        raise ValueError("direction must be 'both', 'in' or 'out'")

    if direction == "both":
        t = np.linspace(-cyc * 2 * np.pi, cyc * 2 * np.pi, 1000)
    elif direction == "out":
        t = np.linspace(0, cyc * 2 * np.pi, 1000)
    else:  # in
        t = np.linspace(0, -cyc * 2 * np.pi, 1000)

    r = (np.cos(np.pi / n)) ** (-n * t / np.pi)
    ang = t + theta + rotation
    x = a * r * np.cos(ang)
    y = b * r * np.sin(ang)

    if not plot:
        return x, y

    trace = go.Scatter(
        x=x,
        y=y,
        mode="lines",
        line=dict(color=color, width=line_width),
        opacity=alpha,
        showlegend=False,
    )
    return trace


def nSpiral(n=4, N=4, cyc=0.5, color="b", theta=0, rotation=0, line_width=1.0):
    """绘制 n 瓣螺旋。"""
    traces = []
    for i in range(N):
        t1 = logSpiral(
            n,
            1,
            1,
            cyc,
            color,
            theta + i * 2 * np.pi / N,
            rotation,
            "both",
            1,
            line_width,
        )
        t2 = logSpiral(
            n,
            -1,
            1,
            cyc,
            color,
            -theta + i * 2 * np.pi / N,
            -rotation,
            "both",
            1,
            line_width,
        )
        if t1:
            traces.append(t1)
        if t2:
            traces.append(t2)
    return traces


def logSpiral_in_out(n, a, b, cyc, color="b", theta=0):
    """绘制双向对数螺线。"""
    t = np.linspace(-cyc * 2 * np.pi, cyc * 2 * np.pi, 1000)
    x = a * (np.cos(np.pi / n)) ** (-n * t / np.pi) * np.cos(t + theta)
    y = b * (np.cos(np.pi / n)) ** (-n * t / np.pi) * np.sin(t + theta)
    trace = go.Scatter(
        x=x, y=y, mode="lines", line=dict(color=color), showlegend=False
    )
    return trace


def logSpiral_out(n, a, b, cyc, color="b", theta=0):
    """绘制外向对数螺线。"""
    t = np.linspace(0, cyc * 2 * np.pi, 100)
    x = a * (np.cos(np.pi / n)) ** (-n * t / np.pi) * np.cos(t + theta)
    y = b * (np.cos(np.pi / n)) ** (-n * t / np.pi) * np.sin(t + theta)
    trace = go.Scatter(
        x=x, y=y, mode="lines", line=dict(color=color), showlegend=False
    )
    return trace


def logSpiral_in(n, a, b, cyc, color="b", theta=0):
    """绘制内向对数螺线。"""
    t = np.linspace(0, -cyc * 2 * np.pi, 100)
    x = a * (np.cos(np.pi / n)) ** (-n * t / np.pi) * np.cos(t + theta)
    y = b * (np.cos(np.pi / n)) ** (-n * t / np.pi) * np.sin(t + theta)
    trace = go.Scatter(
        x=x, y=y, mode="lines", line=dict(color=color), showlegend=False
    )
    return trace


def n_spiral(n, cyc, color, theta=0):
    """绘制 n 个螺旋。"""
    traces = []
    for i in range(n):
        t1 = logSpiral(n, 1, 1, cyc, color, theta + i * 2 * np.pi / n)
        t2 = logSpiral(n, -1, 1, cyc, color, theta + i * 2 * np.pi / n)
        if t1:
            traces.append(t1)
        if t2:
            traces.append(t2)
    return traces


def n_spiral_rotate(n, cyc, color, alpha=0, theta=0):
    """绘制旋转的 n 个螺旋。"""
    traces = []
    for i in range(n):
        t1 = logSpiral(n, 1, 1, cyc, color, alpha + theta + i * 2 * np.pi / n)
        t2 = logSpiral(n, -1, 1, cyc, color, alpha - theta + i * 2 * np.pi / n)
        if t1:
            traces.append(t1)
        if t2:
            traces.append(t2)
    return traces


def n_spiral_rotate_out(n, cyc, color, theta=0):
    """绘制旋转的外向 n 个螺旋。"""
    traces = []
    for i in range(n):
        t1 = logSpiral_out(n, 1, 1, cyc, color, theta + i * 2 * np.pi / n)
        t2 = logSpiral_out(n, -1, 1, cyc, color, -theta + i * 2 * np.pi / n)
        if t1:
            traces.append(t1)
        if t2:
            traces.append(t2)
    return traces


def n_spiral_rotate_in(n, cyc, color, theta=0):
    """绘制旋转的内向 n 个螺旋。"""
    traces = []
    for i in range(n):
        t1 = logSpiral_in(n, 1, 1, cyc, color, theta + i * 2 * np.pi / n)
        t2 = logSpiral_in(n, -1, 1, cyc, color, -theta + i * 2 * np.pi / n)
        if t1:
            traces.append(t1)
        if t2:
            traces.append(t2)
    return traces


def calla_petal(n, cyc, theta, color):
    """绘制马蹄莲花瓣。"""
    traces = []
    t1 = logSpiral(n, 1, 1, cyc * 1.25, color, theta)
    t2 = logSpiral(n, -1, 1, cyc * 1.25, color, -theta)
    if t1:
        traces.append(t1)
    if t2:
        traces.append(t2)
    return traces


def calla_by_petal(n, cyc, N, theta, colors):
    """绘制多个马蹄莲花瓣。"""
    traces = []
    for i in range(N):
        petal_traces = calla_petal(n, cyc, theta + i * 2 * np.pi / N, colors[i])
        traces.extend(petal_traces)
    return traces
